sdk/storage/Azure.Storage.DataMovement/src/UriToStreamTransferJob.cs

```diff
@@ -71,49 +72,159 @@ public override async IAsyncEnumerable<JobPartInternal> ProcessJobToJobPartAsync
+                    catch (Exception ex)
+                    {
+                        await InvokeFailedArgAsync(ex).ConfigureAwait(false);
+                        yield break;
+                    }
                     yield return part;
                 }
                 else
                 {
-                    // Call listing operation on the source container
-                    await foreach (StorageResource resource
-                        in _sourceResourceContainer.GetStorageResourcesAsync(
-                            cancellationToken: _cancellationToken).ConfigureAwait(false))
+                    await foreach (JobPartInternal part in GetStorageResourcesAsync().ConfigureAwait(false))
                     {
-                        // Pass each storage resource found in each list call
-                        string sourceName = resource.Path.Substring(_sourceResourceContainer.Path.Length + 1);
-                        UriToStreamJobPart part = await UriToStreamJobPart.CreateJobPartAsync(
-                            job: this,
-                            partNumber: partNumber,
-                            sourceResource: resource,
-                            destinationResource: _destinationResourceContainer.GetChildStorageResource(sourceName),
-                            length: resource.Length).ConfigureAwait(false);
-                        _jobParts.Add(part);
-
                         yield return part;
-                        partNumber++;
                     }
                 }
             }
             else
             {
                 // Resuming old job with existing job parts
+                bool isFinalPartFound = false;
                 foreach (JobPartInternal part in _jobParts)
                 {
-                    // Skip over job parts that have already completed. If they were in a failed
-                    // or skipped state we can retry them.
                     if (part.JobPartStatus != StorageTransferStatus.Completed)
                     {
                         part.JobPartStatus = StorageTransferStatus.Queued;
                         yield return part;
+
+                        if (part.IsFinalPart)
+                        {
+                            // If we found the final part then we don't have to relist the container.
+                            isFinalPartFound = true;
+                        }
+                    }
+                }
+                if (!isFinalPartFound)
+                {
+                    await foreach (JobPartInternal jobPartInternal in GetStorageResourcesAsync().ConfigureAwait(false))
+                    {
+                        yield return jobPartInternal;
                     }
                 }
             }
             _enumerationComplete = true;
             await OnEnumerationComplete().ConfigureAwait(false);
         }
+
+        private async IAsyncEnumerable<JobPartInternal> GetStorageResourcesAsync()
```

Comment:
This function exists on each job type, can we move it to base class or a separate helper somehow?

Reply:
Sure, no problem. Just probably mark something in that issue so we don't forget :).